mark dead traces

a-arefina opened the pull request:
keeping dead trace indicator in headers allows convenient fitering

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -277,11 +278,17 @@ def collect_stats(self, indices=None, n_quantile_traces=100000, quantile_precisi
             if i < n_quantile_traces:
                 traces_buf[i] = trace
+                
+        self.n_dead_traces = len(dead_indices)
```
Remove `self.n_dead_traces = 0` line above then

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -277,11 +278,17 @@ def collect_stats(self, indices=None, n_quantile_traces=100000, quantile_precisi
+                
+        self.n_dead_traces = len(dead_indices)
+        self.headers['DeadTrace'] = 0
+        self.headers.loc[self.headers["TRACE_SEQUENCE_FILE"].isin(dead_indices), 'DeadTrace'] = 1
```
You know, that `i` in the above `for` loop represents the index of the processed trace. Thus you can store it in `dead_indices` instead of `pos + 1` to use `iloc` here, which is way faster

a-arefina replied:
but `i` is an index in a shuffled array `praces_pos`, it has no connection with the original traces indices in headers :(

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -277,11 +278,17 @@ def collect_stats(self, indices=None, n_quantile_traces=100000, quantile_precisi
                 traces_buf[i] = trace
+                
+        self.n_dead_traces = len(dead_indices)
+        self.headers['DeadTrace'] = 0
```
Don't forget to mention this new header in docs

alexandrkuvaev commented:
Also think of a `remove_dead_traces` method of a `Survey`, which will run `collect_stats` if they are not present and perform filtering

a-arefina commented:
> Also think of a `remove_dead_traces` method of a `Survey`, which will run `collect_stats` if they are not present and perform filtering

I've added a method for removing dead traces but there are some questions to concider

1) Do we need an option to skip quantiles calculation when running `collect_stats` for performance reasons?
2) deleting dead traces changes survey's stats. I've added a flag to recalculate it, but to do it fairly, we need to save parameters of previous calculation process (indices, n_quantile_traces, quantile_precision, stats_limits). In seismiqb they don't include dead traces in the pool for calculating stats, which removes the need for recalculation after dead traces are deleted. 
3) do we need to recalculate stats in `set_limits`? Maybe there are other methods that can chance stats?

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -675,6 +684,34 @@ def _process_limits(self, limits):
+        """
+        self = maybe_copy(self, inplace)  # pylint: disable=self-cls-assignment
+        if not self.has_stats:
+            self.collect_stats()
```
Let's pass all the `kwargs` here

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -675,6 +684,34 @@ def _process_limits(self, limits):
+
+        self.filter(lambda c: c==0, cols='DeadTrace', inplace=True)
+        self.n_dead_traces = 0
+        if recollect_stats:
```
I'd change `collect_stats` to use only non-dead traces for stats calculation (which seems pretty reasonable) and remove this check at all

alexandrkuvaev commented:
> 1. Do we need an option to skip quantiles calculation when running `collect_stats` for performance reasons?

If you pass `kwargs` as I suggest above, you can do it by passing `n_quantile_traces=0`. But quantile calculation is not generally a bottleneck at least when `n_quantile_traces` is not that great.

> 2. deleting dead traces changes survey's stats. I've added a flag to recalculate it, but to do it fairly, we need to save parameters of previous calculation process (indices, n_quantile_traces, quantile_precision, stats_limits). In seismiqb they don't include dead traces in the pool for calculating stats, which removes the need for recalculation after dead traces are deleted.

I'd calculate stats only for non-dead traces.

> 3. do we need to recalculate stats in `set_limits`? Maybe there are other methods that can chance stats?

I've checked the `Survey`, the following methods can possibly change stats:
* `load_first_breaks`
* `filter`
* `set_limits`
* `generate_supergathers`
All of them change either `headers` or `limits`. One possible way to handle it is to introduce a new flag `dirty_stats`, similar to the one in `VelocityCube`, and check it every time the stats are requested, but I'd keep it as is for now since the fix is not that simple.

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -675,6 +691,31 @@ def _process_limits(self, limits):
+        Parameters
+        ----------
+        inplace : bool, optional
+            Whether to remove data inplace or return a new survey instance, by default False
```
```suggestion
            Whether to remove traces inplace or return a new survey instance, by default False
```

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -675,6 +691,31 @@ def _process_limits(self, limits):
+        if not self.has_stats:
+            self.collect_stats(**kwargs)
+
+        self.filter(lambda dt: dt == False, cols='DeadTrace', inplace=True)
```
```suggestion
        self.filter(lambda dt: ~dt, cols='DeadTrace', inplace=True)
```

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -267,21 +271,29 @@ def collect_stats(self, indices=None, n_quantile_traces=100000, quantile_precisi
         trace = np.empty(self.samples_length, dtype=np.float32)
 
+        dead_indices = []
+        counter = 0
```
I'd rename it to something like `cur_quantile_trace_index`

a-arefina replied:
what about `quantile_traces_counter`?

alexandrkuvaev replied:
Up to you, much better than simply `counter` anyway)

alexandrkuvaev commented on seismicpro/src/survey.py:
```diff
@@ -675,6 +691,31 @@ def _process_limits(self, limits):
 
+    def remove_dead_traces(self, inplace=False, **kwargs):
+        """ Removes dead (constant) traces from survey's data.
+        Calls `collect_stats` if `self.has_stats` flag is not set
```
```suggestion
        Calls `collect_stats` if `self.has_stats` flag is not set.
```

a-arefina commented on seismicpro/src/survey.py:
```diff
@@ -267,21 +271,29 @@ def collect_stats(self, indices=None, n_quantile_traces=100000, quantile_precisi
-        for i, pos in tqdm(enumerate(traces_pos), desc=f"Calculating statistics for survey {self.name}",
-                           total=len(traces_pos), disable=not bar):
-            self.load_trace(buf=trace, index=pos, limits=limits, trace_length=self.samples_length)
+        for i in tqdm(shuffled_indices, desc=f"Calculating statistics for survey {self.name}",
+                           total=num_traces, disable=not bar):
```
```suggestion
        for i in tqdm(shuffled_indices, desc=f"Calculating statistics for survey {self.name}",
                      total=num_traces, disable=not bar):
```
